### PR TITLE
cache compact_index responses and fixes

### PR DESCRIPTION
**Pull request by sonalkr132**
`CompactIndex` really needs to take its belonging to a separate class.

Also, when did we unskip remaining tests? I am pretty sure we were skipping more than one and now we are skipping none.

### CODE REVIEW

File: app/models/rubygem.rb [hunk 1]
```diff
@@ -313,32 +351,30 @@ def compact_index_info
...
+      .where("indexed is false and versions.created_at > ?", date)
+      .order("versions.created_at, versions.number, platform")
 
-  def needs_name_validation?
-    new_record? || name_changed?
-  end
+    # Arel don't support yet order on unions :(
+    # https://github.com/rails/arel/issues/98
+    sql = "(#{created_gems.to_sql}) UNION (#{yanked_gems.to_sql}) ORDER BY date"
+    gems = Rubygem.find_by_sql(sql)
```
Comment (sonalkr132) [inline app/models/rubygem.rb]: Using raw sql here *drops* response time of `versions` from: 9072-9965ms to 2766-3182ms (min-max over 10 requests). I will make this change after this PR gets merged.

ENV:
config.cache_classes = true
config.consider_all_requests_local       = false
config.action_controller.perform_caching = false
versions file: one on bundler-api (created_at: 2016-01-22T16:03:56-06:00)

**Comment by sonalkr132** [comment]
This PR also changes:
* `info_checksum` on versions enpoint was nil because `info_checksum` is fourth argument of `CompactIndex::GemVersion.new` ([Reference](https://github.com/bundler/compact_index/blob/master/lib/compact_index/gem_version.rb#L3)). It is okay to pass `nil` as third argument(happens to be `checksum/sha256`) because `versions` endpoint doesn't use `checksum` ([Referene](https://github.com/bundler/bundler/blob/master/lib/bundler/vendor/compact_index_client/lib/compact_index_client/cache.rb#L24)).
* Use raw sql instead of activerecord query in `compact_index_versions` because it is considerably faster. Checkout:  https://github.com/rubygems/rubygems.org/pull/1327#discussion_r67112773
* Assign default name for unregistered gem using [coalesce](https://www.postgresql.org/docs/9.5/static/functions-conditional.html#FUNCTIONS-COALESCE-NVL-IFNULL) which is later used to filter deps. Checkout https://github.com/rubygems/rubygems.org/pull/1300#discussion_r66801004 for description of problem it solves.
* Update of info_checksum is moved to `Pusher`. Checkout commit message or https://github.com/rubygems/rubygems.org/pull/1300#discussion_r67534937 for reasoning.
* Show md5 format of checksum on `info/:name` endpoint

**Comment by sonalkr132** [comment]
ping! @fotanus

**Comment by indirect** [comment]
@dwradcliffe can we deploy this to staging so we can try it out?

**Comment by indirect** [comment]
@sonalkr132 I have pushed [a few commits](https://github.com/rubygems/rubygems.org/compare/c28ab52...aa-failing-info-test) that include renaming some things for clarity and a failing test for `GET /info/gemC` when `gemC` only has development dependencies.

You can see the correct, expected output by commenting out line 121—but we need the test to pass even when a gem has only development dependencies.

**Comment by indirect** [comment]
Added a few more commits to [my branch](https://github.com/rubygems/rubygems.org/compare/c28ab52...aa-failing-info-test) adding tests for the correct ETag headers, and fixing the ETag headers for all responses except the already failing test.

**Comment by sonalkr132** [comment]
Thanks for looking into it @indirect 
I have fixed the failing test.

**Comment by indirect** [comment]
@sonalkr132 great, thanks! Unfortunately, a long time ago `dependencies.scope` was not required. Because of that, we should add a migration something like `Dependency.where("scope is null").update_all(scope: "runtime")`. Then we can be sure he query will work.

**Comment by sonalkr132** [comment]
> we should add a migration something like Dependency.where("scope is null").update_all(scope: "runtime")

Latest data dump of rubygems.org is available at https://rubygems.org/pages/data
Following is what I get when I look for `null` scope on dependencies table:
```
rubygem_new=# select count(*) from dependencies;
  count  
---------
 3638968
rubygem_new=# select count(*) from dependencies where scope='runtime';
  count  
---------
 1848876
rubygem_new=# select count(*) from dependencies where scope='development';
  count  
---------
 1790092
rubygem_new=# select count(*) from dependencies where scope is null;
 count 
-------
     0
```

I don't think we need the migration?

**Comment by indirect** [comment]
Oh, great! We must have done that migration already. 😄

File: test/integration/api/compact_index_test.rb [hunk 2]
```diff
@@ -2,107 +2,153 @@
...
 
-  test "/info with unexisting gem" do
+  test "/info with nonexistent gem" do
     get info_path(gem_name: 'donotexist')
     assert_response :not_found
-  end
-
-  test "/info partial response" do
-    get info_path(gem_name: @rubygem.name), nil, range: "bytes=159-"
-    assert_response 206
```
Comment (fotanus) [inline test/integration/api/compact_index_test.rb]: I like the idea of keep the `assert_response 206` in here.

Comment (sonalkr132) [inline test/integration/api/compact_index_test.rb]: @fotanus It is not removed, it is moved up [here](https://github.com/rubygems/rubygems.org/pull/1327/files#diff-a708a123bd49a50ac54c92bdcdc6028dR129).
